### PR TITLE
feat: allEqual method

### PR DESCRIPTION
**Pull request by jackmellis**
assert that all array items are equal

### CODE REVIEW

**Comment by guillaumearm** [comment]
I propose a more flexible version :

**e.g.** : 
```js
const allEqual = R.curry((x, arr) => (
  R.all(R.equals(x), arr)
));

allEqual(42, [42, 42, 42]) // => true
allEqual(21, [42, 42, 42]) // => false
```

**Comment by jackmellis** [comment]
The purpose of this method was so that we didn't have to pass in a value to compare to, we just want to check that all of the values are the same, regardless of what that value happens to be.

**Comment by Undistraction** [comment]
I think it's worth adding. Might also be worth adding an `allUnique` that checks there are no duplicates.

If @char0n is happy with this addition I'll review.

**Comment by guillaumearm** [comment]
@Undistraction : `allUnique`is not really the complement, the complement is `notAllEqual`.

Effectively, we don't want to add a `notAllEqualTo` because using `notAllEqual` is enough.

Anyway, maybe we would add a complement of `allUnique` too.

**Comment by Undistraction** [comment]
@guillaumearm You are correct. So that would mean:

- `allEqual` (all values are equal)
- `notAllEqual` (at least one value is different from the others)
- `allUnique` (all values are unique)
- `notAllUnique` (at least two values are identical). 

I wonder if there is a better word than `equal` to use in the naming? Unique captures it perfectly, but 'equal' poses the question 'equal to what?'.

**Comment by guillaumearm** [comment]
> but 'equal' poses the question 'equal to what?'.

This is exactly the point I thought when I proposed `allEqualTo`, maybe @char0n will have an idea.

**Comment by Undistraction** [comment]
I think 'allEqual' is perfectly valid but it needs to be called `allIdentical` or something similar.

**Comment by guillaumearm** [comment]
Yup, I think the same thing but `allIdentical` is the best name IMO.

File: src/allEqual.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { uniq, length, lte, __, pipe } from 'ramda';
+
+/* eslint-disable */
+/**
+  * Checks that all elemnets in a list are the same
+  *
+  *
```
Comment (char0n) [inline src/allEqual.js]: Drop to redundant new lines pls

File: src/allEqual.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { uniq, length, lte, __, pipe } from 'ramda';
+
+/* eslint-disable */
```
Comment (char0n) [inline src/allEqual.js]: Why disable linting ?

File: test/allEqual.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { assert } from 'chai';
+
+import * as RA from '../src/index';
+
+describe('allEqual', function() {
+  it('returns true when all items are equal', function() {
+    assert.equal(RA.allEqual([4, 4, 4, 4]), true);
+  });
+  it('returns true when all items are deeply equal', function() {
```
Comment (char0n) [inline test/allEqual.js]: Use `should` convention in tests. Can be seen here: https://github.com/char0n/ramda-adjunct/blob/f7d7f7c454c2e49806b3fc0e1e521577a38ea544/test/argsPass.js#L32

File: src/allEqual.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+/**
+  * Checks that all elemnets in a list are the same
+  *
+  *
+  *
+  * @func allEqual
+  * @memberOf RA
+  * @category List
+  * @sig [Iterable a] -> boolean
```
Comment (char0n) [inline src/allEqual.js]: Boolean is a category
```js
@sig [Iterable a] -> Boolean
```

File: src/allEqual.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import { uniq, length, lte, __, pipe } from 'ramda';
+
+/* eslint-disable */
+/**
+  * Checks that all elemnets in a list are the same
```
Comment (char0n) [inline src/allEqual.js]: Returns true if all elements of the list match the each other, false if there are any that don't.

File: src/allEqual.js (hunk 4 above)
Comment (char0n) [inline src/allEqual.js]: Add
```js
@since {@link https://char0n.github.io/ramda-adjunct/2.8.0|v2.8.0}

File: src/allEqual.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  *
+  *
+  *
+  * @func allEqual
+  * @memberOf RA
+  * @category List
+  * @sig [Iterable a] -> boolean
+  * @params {Iterable.<*>}
```
Comment (char0n) [inline src/allEqual.js]:
```js
@param {Iterable.<*>} iterable

File: src/allEqual.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
```
Comment (char0n) [inline src/allEqual.js]: Add
```js
@return {boolean}

File: src/allEqual.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  * @example
+  *
+  * allEqual([ 1, 2, 3, 4 ]); //=> false
+  * allEqual([ 1, 1, 1, 1 ]); //=> true
+  */
+/* eslint-enable */
+const allEqual = pipe(uniq, length, lte(__, 1));
```
Comment (char0n) [inline src/allEqual.js]: Prefer
```js
const allEqual = curry((x, arr) => (
  all(equals(x), arr)
));
```

**Comment by char0n** [comment]
`allIdentical` will not be correct name, because we're doing deep equal here. I imagine identity to be different than equal.

File: src/allEqual.js (hunk 8 above)
Comment (guillaumearm) [reply src/allEqual.js]: It's not the same thing, this implementation had arity 2 and should be named `allEqualTo`

**Comment by guillaumearm** [comment]
@char0n, you're totally right, we have the same problem with `allUnique` IMO, it suggests that all values are unique in term of references.

anyway, we should back to the initial proposed `allEqual` name, it suggests : `all equal to each other`.

File: src/allEqual.js (hunk 8 above)
Comment (Undistraction) [reply src/allEqual.js]: I agree with @guillaumearm. I think this is a separate function: `allEqualTo`.

**Comment by Undistraction** [comment]
@guillaumearm I don't think 'unique' implies that. `R.uniq` is using `R.equals` for comparison - so in Ramda terms 'unique' means values that are not deep-equal. So I think we can legitimately use 'allUnique' (or possibly `allUniq` for consistency.

**Comment by Undistraction** [comment]
This is a tough one as `equals` is verbose and `uniq` is contracted. But given that this is a sister function to `allEquals` I vote for `allUnique`, especially as the use of `R.uniq` is actually an implementation detail.

**Comment by guillaumearm** [comment]
So, I'm a bit confused for this PR, It's `allEqual` or `allEquals` ?

Let's talk about `AllUnique` in the concerned thread.

**Comment by Undistraction** [comment]
As far as I understand it:

1. This PR just needs to be completed as per @charOn's review. Current name is correct (allEqual).

2. We need to vote for naming of allUniq Vs allUnique (I vote for the latter).

3. We need to decide if we want allEqualTo which takes an additional value to compare values to.

**Comment by guillaumearm** [comment]
2. Let's vote in #515 
3. I just created #525

File: src/allEqual.js (hunk 8 above)
Comment (char0n) [reply src/allEqual.js]: Agree. I missed the previous comments. Removing this comment.

**Comment by char0n** [comment]
@Undistraction, @guillaumearm as this haven't been touched for longer time, there are comuple of options what we can do here:

1.) @jackmellis will  continue with the PR and finishes it (@jackmellis is that something you're willing to do?)
2.) we close this PR and create a new one with the same functionality
3.) we close it

**Comment by jackmellis** [comment]
I'm quite happy to fix up the PR for the `allEqual` method - just been a bit busy with work lately.

**Comment by char0n** [comment]
Have some usecases for this so I managed to continue with it here: https://github.com/char0n/ramda-adjunct/pull/572

**Comment by Undistraction** [comment]
I'm going to go ahead and close as this is now added by #513

**Pull request closed without merging**